test/common/services/security/test_user.ts

```diff
@@ -0,0 +1,96 @@
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+import { Role } from './role';
+import { User } from './user';
+import { FtrProviderContext } from '../../ftr_provider_context';
+import { Browser } from '../../../functional/services/browser';
+import { TestSubjects } from '../../../functional/services/test_subjects';
+
+export async function createTestUserService(
+  role: Role,
+  user: User,
+  { getService, hasService }: FtrProviderContext
+) {
+  const log = getService('log');
+  const config = getService('config');
+  // @ts-ignore browser service is not normally available in common.
+  const browser: Browser | void = hasService('browser') && getService('browser');
+  const testSubjects: TestSubjects | void =
+    // @ts-ignore testSubject service is not normally available in common.
+    hasService('testSubjects') && getService('testSubjects');
+  const kibanaServer = getService('kibanaServer');
+
+  const enabledPlugins = config.get('security.disableTestUser')
+    ? []
+    : await kibanaServer.plugins.getEnabledIds();
+  const isEnabled = () => {
+    return enabledPlugins.includes('security') && !config.get('security.disableTestUser');
+  };
+  if (isEnabled()) {
+    log.debug(
+      '====================================================creating roles and users========================='
+    );
+    for (const [name, definition] of Object.entries(config.get('security.roles'))) {
+      // create the defined roles (need to map array to create roles)
+      await role.create(name, definition);
+    }
+    try {
+      // delete the test_user if present (will it error if the user doesn't exist?)
+      await user.delete('test_user');
+    } catch (exception) {
+      log.debug('no test user to delete');
+    }
+
+    // create test_user with username and pwd
+    log.debug(`default roles = ${config.get('security.defaultRoles')}`);
+    await user.create('test_user', {
+      password: 'changeme',
+      roles: config.get('security.defaultRoles'),
+      full_name: 'test user',
+    });
+  }
+
+  return new (class TestUser {
+    async restoreDefaults() {
+      this.assertIsEnabled();
+      await this.setRoles(config.get('security.defaultRoles'));
+    }
+
+    async setRoles(roles: string[]) {
+      this.assertIsEnabled();
+      log.debug(`set roles = ${roles}`);
+      await user.create('test_user', {
+        password: 'changeme',
+        roles,
+        full_name: 'test user',
+      });
+
+      if (browser && testSubjects) {
+        await browser.refresh();
+        await testSubjects.find('kibanaChrome', config.get('timeouts.find') * 10);
+      }
```

Comment:
I'm not sure about mixing API calls with Browser interactions under `createTestUserService`:

This way it limits us to execute `setRoles` only when webdriver session is started, but I guess we may later decide to use it in non-Selenium tests?
I would rather add a function in CommonPage that wraps up `setRoles` and browser refresh (since we no POs are used only for Selenium tests)

Reply:
I don't think this limits us because we're using `hasService('browser')` to detect when the browser service is defined, which indicates that we can call `browser.refresh()`. Going this route prevents us from having multiple implementations of this method and makes it clearer to users which method they should be using too.